cron.php

```diff
@@ -0,0 +1,464 @@
+					// There are some jobs we never care to run
+					if ( $this->is_blocked_event( $action ) ) {
+						wp_unschedule_event( $timestamp, $action, $instance_args['args'] );
+						continue;
+					}
+
+					// Queue internal events separately to avoid them being blocked
+					$queue = $this->is_internal_event( $action ) ? 'internal_events' : 'current_events';
+
+					array_push( $$queue, array(
+						'timestamp' => $timestamp,
+						'action'    => md5( $action ),
+						'instance'  => $instance,
+					) );
+				}
+			}
+		}
+
+		// Limit batch size to avoid resource exhaustion
+		if ( count( $current_events ) > $this->job_queue_size ) {
+			$current_events = array_slice( $current_events, 0, $this->job_queue_size );
+		}
+
+		return rest_ensure_response( array(
+			'events'          => $current_events,
+			'internal_events' => $internal_events,
+			'endpoint'        => get_rest_url( null, $this->namespace . '/event/' ),
+		) );
+	}
+
+	/**
+	 * Execute a specific event
+	 */
+	public function run_event( $request ) {
+		// Parse request
+		$event     = $request->get_json_params();
+		$timestamp = isset( $event['timestamp'] ) ? absint( $event['timestamp'] ) : null;
+		$action    = isset( $event['action'] ) ? trim( sanitize_text_field( $event['action'] ) ) : null;
+		$instance  = isset( $event['instance'] ) ? trim( sanitize_text_field( $event['instance'] ) ) : null;
+		unset( $event );
+
+		// Validate input data
+		if ( empty( $timestamp ) || empty( $action ) || empty( $instance ) ) {
+			return new WP_Error( 'missing-data', __( 'Invalid or incomplete request data.', 'wp-cron-control-revisited' ) );
+		}
+
+		// Ensure we don't run jobs too far ahead
+		if ( $timestamp > strtotime( sprintf( '+%d seconds', $this->job_execution_buffer_in_seconds ) ) ) {
+			return new WP_Error( 'premature', __( 'Event is not scheduled to be run yet.', 'wp-cron-control-revisited' ) );
+		}
+
+		// Find the event to retrieve the full arguments
+		$event = $this->get_event( $timestamp, $action, $instance );
+		unset( $timestamp, $action, $instance );
+
+		if ( is_array( $event ) ) {
+			$time_start = microtime( true );
+
+			// Limit how many events are processed concurrently
+			if ( ! $this->is_internal_event( $event['action'] ) && ! $this->check_lock() ) {
+				return new WP_Error( 'no-free-threads', __( 'No resources available to run this job.', 'wp-cron-control-revisited' ) );
+			}
+
+			// Prepare environment to run job
+			ignore_user_abort( true );
+			set_time_limit( $this->job_timeout_in_minutes * MINUTE_IN_SECONDS );
+			define( 'DOING_CRON', true );
+
+			// Remove the event, and reschedule if desired
+			// Follows pattern Core uses in wp-cron.php
+			if ( false !== $event['schedule'] ) {
+				$reschedule_args = array( $event['timestamp'], $event['schedule'], $event['action'], $event['args'] );
+				call_user_func_array( 'wp_reschedule_event', $reschedule_args );
+			}
+
+			wp_unschedule_event( $event['timestamp'], $event['action'], $event['args'] );
```

Comment:
I wonder if there's a possible race condition here: two jobs trigger at the exact same time and which means that `wp_unschedule_event` for both is triggered at the same time for both as well. One of them will win and likely overwrite the other (since all cron is stored in a single option entry)?